ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+
+       The following (seemingly redundant) several lines prevents that:
+       - If a <= (f v).lower, immediately succeed
+       - If not (a <= (f v).upper), immediately fail
+       - Note that at this point, we still can't ensure that a >= (f v).lower.
+         (We don't assume total ordering for generality.)
+        Therefore, we set a = join a (f v).lower. This operation has no effect
+        in terms of submoding, but we have now ensured that a is within the
+        range of f v, and thus a is within the co-domain of f, and thus its
+        safe to apply f' to a. *)
+    if C.le obj a (mlower obj mv)
+    then Ok ()
+    else if not (C.le obj a (mupper obj mv))
+    then Error (mupper obj mv)
+    else
+      let a = C.join obj a (mlower obj mv) in
+      let f' = C.left_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      assert (Result.is_ok (submode_cv ?log src a' v));
+      Ok ()
+
+  (* In the following, functions might take two logs: [log] records changes
+     caused by the operation, [optim] records optimizing changes not caused by the
+     operation. Semantically, [optim] happens before [log] and doesn't rely on [log].
+     On the other hand, some changes in [log] might rely on [optim]. *)
+
+  (** Returns [None] if success; [Some x] if failed, and [x] is the next best
+     guess to replace the constant argument that MIGHT succeed. *)
+  let rec submode_vc :
+      type a. ?log:_ -> a C.obj -> a var -> a -> (unit, a) Result.t =
+    fun (type a) ?log (obj : a C.obj) v a' ->
+     if C.le obj v.upper a'
+     then Ok ()
+     else if not (C.le obj v.lower a')
+     then Error v.lower
+     else (
+       update_upper ?log obj v a';
+       let r =
+         v.vlower
+         |> find_error (fun mu ->
+                let r = submode_mvc ?log obj mu a' in
+                (if Result.is_ok r
+                then
+                  (* Optimization: update [v.lower] based on [mlower u].*)
+                  let mu_lower = mlower obj mu in
+                  if not (C.le obj mu_lower v.lower)
+                  then update_lower ?log obj v mu_lower);
+                r)
+       in
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
+       r)
+
+  and submode_mvc :
+        'a 'r.
+        ?log:change list ref ->
+        'a C.obj ->
+        ('a, allowed * 'r) morphvar ->
+        'a ->
+        (unit, 'a) Result.t =
+   fun ?log obj (Amorphvar (v, f) as mv) a ->
+    (* See [submode_cmv] for why we need the following seemingly redundant
+       lines. *)
+    if C.le obj (mupper obj mv) a
+    then Ok ()
+    else if not (C.le obj (mlower obj mv) a)
+    then Error (mlower obj mv)
+    else
+      let a = C.meet obj a (mupper obj mv) in
+      let f' = C.right_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      (* cannot assert [Option.is_none] because [mlower] is imprecise. *)
```

Comment:
If `mlower` was precise, then L321 should ensures `f v <= a`, and thus L329 should never fail, and we would be able to assert that.

We are able to assert that in `submode_cmv`, because `mupper` is precise.

(This is extra confusing, because `Option` is now replaced with `Result`. I will fix)